File: RemoteInstaller/app/src/main/java/remoteapkinstallerclient/com/br/remoteinstaller/view/AppsAdapter.java
package remoteapkinstallerclient.com.br.remoteinstaller.view;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Date;
import java.util.List;

import remoteapkinstallerclient.com.br.remoteinstaller.R;
import remoteapkinstallerclient.com.br.remoteinstaller.objects.App;
import remoteapkinstallerclient.com.br.remoteinstaller.service.DownloadService;
import remoteapkinstallerclient.com.br.remoteinstaller.service.DownloadTask;
import remoteapkinstallerclient.com.br.remoteinstaller.service.Urls;
import remoteapkinstallerclient.com.br.remoteinstaller.utils.DateUtils;

/**
 * Created by joao.roriz on 04/12/17.
 */

public class AppsAdapter extends  RecyclerView.Adapter<AppsAdapter.AppsViewHolder>{
    private List<App> apps;
    private Context context;

    public AppsAdapter(List<App> apps, Context context) {
        this.apps = apps;
        this.context = context;
    }

    @Override
    public AppsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_app, parent, false);
        return  new AppsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(AppsViewHolder holder, int position) {
        final App app = apps.get(position);

        holder.tvName.setText(app.getName());
        holder.tvHour.setText(DateUtils.toString("dd/MM/yyyy",new Date(app.getAddedDate())) +" - "+ DateUtils.toString("HH:mm", new Date(app.getAddedDate()) ));
        holder.tvPackage.setText(app.getPackageName());

        holder.btnInstall.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ProgressDialog mProgressDialog;

// instantiate it within the onCreate method
//                mProgressDialog = new ProgressDialog(context);
//                mProgressDialog.setMessage("Aguarde");
//                mProgressDialog.setIndeterminate(true);
//                mProgressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
//                mProgressDialog.setCancelable(true);
//
//                final DownloadTask downloadTask = new DownloadTask(context);
//                downloadTask.execute(Urls.DOWNLOAD_APP_BASE_URL + URLEncoder.encode(app.getChecksum()), "UTF-8");
//
//                mProgressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
//                    @Override
//                    public void onCancel(DialogInterface dialog) {
//                        downloadTask.cancel(true);
//                    }
//                });
                try {
                    new DownloadService().download(context,app);
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
            }
        });

        holder.btnChangelogs.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AlertDialog.Builder alertDialog = new AlertDialog.Builder(context);
                alertDialog.setTitle(app.getName());

                String message ="";
                if(app.getChangelog() != null && app.getChangelog().isEmpty()){
                    message = "Nenhum changlog registrado";
                }else{
                    for (String log : app.getChangelog()){
                        message = message + log+"\n";
                    }
                }
                alertDialog.setMessage(message)
                        .setCancelable(false)
                        .setPositiveButton("Ok",new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog,int id) {
                                dialog.cancel();
                            }
                        });

                alertDialog.show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return apps.size();
    }

    class AppsViewHolder extends RecyclerView.ViewHolder{
        TextView tvName;
        TextView tvPackage;
        TextView tvHour;
        Button btnChangelogs;
        Button btnInstall;

        public AppsViewHolder(View itemView) {
            super(itemView);

            tvName = itemView.findViewById(R.id.tv_nome);
            tvHour = itemView.findViewById(R.id.tv_hour);
            tvPackage = itemView.findViewById(R.id.tv_app_package);
            btnChangelogs = itemView.findViewById(R.id.btn_changelos);
            btnInstall = itemView.findViewById(R.id.btn_install);

        }
    }
}

File: RemoteInstaller/app/src/main/java/remoteapkinstallerclient/com/br/remoteinstaller/view/adapter/AppsAdapter.kt
package remoteapkinstallerclient.com.br.remoteinstaller.view.adapter

import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView

import java.io.UnsupportedEncodingException
import java.util.Date

import remoteapkinstallerclient.com.br.remoteinstaller.R
import remoteapkinstallerclient.com.br.remoteinstaller.objects.App
import remoteapkinstallerclient.com.br.remoteinstaller.service.download.DownloadService
import remoteapkinstallerclient.com.br.remoteinstaller.utils.DateUtils

/**
 * Created by joao.roriz on 04/12/17.
 */

class AppsAdapter(private val apps: List<App>?, private val context: Context) : RecyclerView.Adapter<AppsAdapter.AppsViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppsViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_app, parent, false)
        return AppsViewHolder(view)
    }

    override fun onBindViewHolder(holder: AppsViewHolder, position: Int) {
        val app = apps!![position]

        holder.tvName.text = app.appName
        holder.tvHour.text = DateUtils.toString("dd/MM/yyyy", Date(app.addedDate!!)) + " - " + DateUtils.toString("HH:mm", Date(app.addedDate!!))
        holder.tvPackage.text = app.packageName

        holder.btnInstall.setOnClickListener {
            val mProgressDialog: ProgressDialog
            try {
                DownloadService().download(context, app)
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
            }
        }

        holder.btnChangelogs.setOnClickListener {
            val alertDialog = AlertDialog.Builder(context)
            alertDialog.setTitle(app.appName)

            var message = ""
            if (app.changelog != null && app.changelog!!.isEmpty()) {
                message = "Nenhum changlog registrado"
            } else {
                for (log in app.changelog!!) {
                    message = message + log + "\n"
                }
            }
            alertDialog.setMessage(message)
                    .setCancelable(false)
                    .setPositiveButton("Ok") { dialog, id -> dialog.cancel() }

            alertDialog.show()
        }
    }

    override fun getItemCount(): Int {
        return apps?.size ?: 0
    }

    class AppsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var tvName: TextView
        var tvPackage: TextView
        var tvHour: TextView
        var btnChangelogs: Button
        var btnInstall: Button

        init {

            tvName = itemView.findViewById(R.id.tv_nome)
            tvHour = itemView.findViewById(R.id.tv_hour)
            tvPackage = itemView.findViewById(R.id.tv_app_package)
            btnChangelogs = itemView.findViewById(R.id.btn_changelos)
            btnInstall = itemView.findViewById(R.id.btn_install)

        }
    }
}
